prepare_check_cleanup/migration.py

```diff
@@ -0,0 +1,269 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utility for running the migration action, primarily for testing purposes."""
+
+import argparse
+import json
+import logging
+import os
+import sys
+from contextlib import suppress
+from enum import Enum
+from pathlib import Path
+
+from github.GithubException import GithubException
+from github.GitRef import GitRef
+from github.PullRequest import PullRequest
+from github.Repository import Repository
+
+from prepare_check_cleanup import exit_
+from src.discourse import create_discourse
+from src.exceptions import DiscourseError
+from src.index import DOCUMENTATION_FOLDER_NAME, DOCUMENTATION_INDEX_FILENAME
+from src.pull_request import (
+    ACTIONS_PULL_REQUEST_TITLE,
+    DEFAULT_BRANCH_NAME,
+    create_repository_client,
+)
+from src.reconcile import NAVIGATION_TABLE_START
+
+
+class Action(str, Enum):
+    """The actions the utility can take.
+
+    Attrs:
+        PREPARE: Prepare discourse pages before running the migration.
+        CHECK_BRANCH: Check that the migration branch was created.
+        CHECK_PULL_REQUEST: Check that the migration pull request was created.
+        CLEANUP: Delete discourse pages before after the migration.
+    """
+
+    PREPARE = "prepare"
+    CHECK_BRANCH = "check-branch"
+    CHECK_PULL_REQUEST = "check-pull-request"
+    CLEANUP = "cleanup"
+
+
+def main() -> None:
+    """Execute requested migration action.
+
+    Raises:
+        NotImplementedError: if an action was received for which there is no imlpementation.
+    """
+    logging.basicConfig(level=logging.INFO)
+
+    parser = argparse.ArgumentParser(
+        prog="MigrationTestSupport",
+        description="Support functions for the migration testing.",
+    )
+    # Consolidating this code should be done in a following pull request with reconcile.py
+    # pylint: disable=duplicate-code
+    parser.add_argument(
+        "--action", help="Action to run", choices=tuple(action.value for action in Action)
+    )
+    parser.add_argument(
+        "--action-kwargs", help="Arguments for the action as a JSON mapping", default="{}"
+    )
+    args = parser.parse_args()
+    # pylint: enable=duplicate-code
+    action_kwargs = json.loads(args.action_kwargs)
+
+    match args.action:
+        case Action.PREPARE.value:
+            prepare(**action_kwargs)
+            sys.exit(0)
+        case Action.CHECK_BRANCH.value:
+            exit_.with_result(check_branch(**action_kwargs))
+        case Action.CHECK_PULL_REQUEST.value:
+            exit_.with_result(check_pull_request(**action_kwargs))
+        case Action.CLEANUP.value:
+            cleanup(**action_kwargs)
+            sys.exit(0)
```

Comment:
No validation of kwargs done before passing them to the action methods ?
Would it be interesting ?

Reply:
This is just a CLI to support the e2e testing. If the arguments are badly formed, it will fail which will be obvious in CI. If this was intended for use outside of this repo we should do that for sure. This CLI is also not covered by any tests for the same reason